client.go

```diff
@@ -1,137 +1,185 @@
-package main
+// Package iconik API for Golang
+package iconik
 
 import (
 	"bytes"
 	"encoding/json"
-	"flag"
-	"net/url"
-
-	"fmt"
+	"io"
 	"io/ioutil"
 	"log"
 	"net/http"
-	"os"
-)
-
-var (
-	tag = "GPTeaching"
+	"sync"
 )
 
 const (
-	searchUrl = "https://app.iconik.io/API/search/v1/search/"
+	iconikHost = "https://app.iconik.io/API/"
+	v1         = "/search/v1/"
 )
 
-// JSON Object structs.
-
-type SearchCriteriaSchema struct {
-	DocTypes []string     `json:"doc_types"`
-	Filter   SearchFilter `json:"filter"`
+// Credentials are the identification required by the Iconik API
+//
+// The AppID is the application key id that you
+// get when generating an application key.
+//
+// The Token is the string generated alongside the
+// application key.
+type Credentials struct {
+	AppID string
+	Token string
 }
 
-type SearchFilter struct {
-	Operator string       `json:"operator"`
-	Terms    []FilterTerm `json:"terms"`
-}
+// IClient implements a Iconik client. Do not modify state concurrently.
+type IClient struct {
+	Credentials
 
-type FilterTerm struct {
-	Name  string `json:"name"`
-	Value string `json:"value"`
-}
+	// If true, don't retry requests if authorization has expired
+	NoRetry bool
 
-type SearchResponse struct {
-	Objects []IconikObject `json:"objects"`
-}
+	// If true, display debugging information about API calls
+	Debug bool
 
-type IconikObject struct {
-	Files []IconikFile `json:"files"`
+	// State
+	mutex      sync.Mutex
+	host       string
+	httpClient http.Client
 }
 
-type IconikFile struct {
-	Name string `json:"name"`
+// NewIClient creates a new Client for accessing the Iconik API.
+func NewIClient(creds Credentials, host string) (*IClient, error) {
+	if host == "" {
+		host = iconikHost
+	}
+	c := &IClient{
+		Credentials: creds,
+		host:        iconikHost,
+	}
+
+	return c, nil
 }
 
-// Iconik API Client
+// Create an authorized request using the client's credentials
+func (c *IClient) authRequest(method, apiPath string, body io.Reader, header http.Header) (*http.Request, error) {
```

Comment:
seems like a very small function call, how about we just make a single function/method called NewRequest that will create the path and the headers and then the http.NewRequest.

Reply:
Changed to newRequest and added header creation!